### DISCUSSION TITLE
[C++] Segfault accessing element of SX matrix

### DISCUSSION THREAD

**Discussion by mbasescu** [casadi/casadi#3075] [General]
@jaeandersson 
Link to previous discussion: https://groups.google.com/g/casadi-users/c/csX2y2TaRQw

As an update to this, I think I've finally identified the problem. The larger context in which the error was occurring is as follows:

A high level object (let's call it `OptimizerCollection`, executing in thread `0`) owns two custom `ThreadedOptimizer` objects, each with their own set of symbolic expressions and `nlpsol` objects; let's call them `threaded_optimizer_1` and `threaded_optimizer_2`. These `ThreadedOptimizer` objects are constructed sequentially in thread `0`. Throughout the execution of the program, thread `0` will sometimes make a call to one of the `ThreadedOptimizer`s that results in a thread being spawned off internal to that optimizer to evaluate its `nlpsol` object (i.e. perform the optimization). Let's examine `threaded_optimizer_1` specifically, and call the thread that it creates thread `1`. We do not construct any new symbolic expressions in any threads while thread `1` is active, and do not make any calls to `threaded_optimizer_2` during this time either. If thread `1` completes, we may at some point make a similar call to `threaded_optimizer_2` resulting in a thread `2` being created and eventually rejoined, with similar rules. Basically, only one optimization problem is ever being constructed or run at any given time.

What causes the memory corruption (although not completely deterministic) is when `OptimizerCollection` is destroyed, but the `threaded_optimizer_1` object is actively performing an optimization. If a `ThreadedOptimizer`'s destructor is called, it first blocks until any owned threads have been completed. However, because of my declaration order in `OptimizerCollection`, `threaded_optimizer_2` is destroyed first and appears to sometimes cause memory corruption in `threaded_optimizer_1`, even though all memory that I control directly in my application is copied by value into thread `1`.

My belief is that there might be some nasty thread-safety issues with the CasADi `SharedObject`/`SharedObjectInternal` objects, and perhaps in other places. Although the `Sparsity` objects in `Sparsity:getScalar()` and friends are static, a shallow copy of the Sparsity object is created when you do `*this = getScalar()` in `assignCached()`. This results in some reference counting getting triggered since `Sparsity` inherits from `SharedObject`. If another object which owns the same `SharedObjectInternal` object happens to be doing reference counting simultaneously in a different thread, you could potentially call `count_down()` twice back to back before calling `count_up()` to say "hey, I still own this thing" - this would cause the `SharedObjectInternal` to get deleted if there were no other references.

I can work around this issue for my case by preventing CasADi from getting called in a multi-threaded fashion (since I was already doing that for the most part to avoid some MUMPS thread safety issues).

Do you agree with this assessment? And is this an intended use-case in which CasADi is expected to be threadsafe?

**Comment by mbasescu** [comment]
I apologize that I don't have the time right now to produce an isolated example, but basically the way I was able to trigger it in my application is to create a situation where a reference to a CasADi SharedObjectInternal gets destroyed while the sole other reference is being used / reference counted in a separate thread. Looping this enough times (thousands) eventually produced a segfault for me.

**Comment by jaeandersson** [comment]
I see. No, the CasADi symbolics and Sparsity class are *not* generally thread-safe during construction, only in evaluation. I am sure that's written somewhere, but maybe it should be made clearer, for example in the FAQ

**Comment by mbasescu** [comment]
Gotcha - no worries. In retrospect I had seen some of the threads discussing CasADi thread safety before (e.g. https://sourceforge.net/p/casadi/discussion/1271244/thread/571ecad7/), I just didn't understand the full picture that there were static variables being copied that have pointers to (non-threadsafe) reference counted dynamic memory. It also could have been that I was just doing some selective reading from that post:

"CasADi is not thread safe during symbolic calculations if the different threads share symbolic expressions."
"So if you can afford constructing your SXFunction instances one at a time, it should hopefully work."
which are conditions that I thought I'm satisfying, but missing:
"Note that constants such as 0, 1, -1 are defined as static, so it's hard to write expressions that do not have common subexpressions."

Out of curiosity, are you using static to optimize for memory usage or is it a different aspect of performance you're targeting? I didn't notice much performance difference for my problems when I experimented with removing those static Sparsity objects and just created new instances every time.

Either way, thanks for the support!

**Reply by jaeandersson** [reply]
> Out of curiosity, are you using static to optimize for memory usage or is it a different aspect of performance you're targeting? I didn't notice much performance difference for my problems when I experimented with removing those static Sparsity objects and just created new instances every time.

Mainly it's about avoiding to creates duplicates of the same expression. Constants such as 0, 1, -1 are used a lot, e.g. in derivatives rules, so you don't want to have each instance correspond to a different class instantiation.

It's tricky to replace the static in this case. You could have each constant keep a reference to some class it belongs to, which would be a pretty major refactoring of CasADi which would mean a lot of work for us without benefitting many users. An easier fix is to ensure that no more than one thread enters a non-thread safe region at a time.
